### PR TITLE
feat(rome_js_analyze): `noExtraLabels`

### LINKED ISSUE #4125: `noExtraLabels`, no-extra-labels
### Description

https://eslint.org/docs/latest/rules/no-extra-label

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

Closes #4125

## Test Plan

_ESLint_'s unit tests included.

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

File: crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    fn insert(&mut self, label: String, breakable_id: BreakableId) {
+        self.labels.insert((self.root_id, label), breakable_id);
+    }
+
+    fn remove(&mut self, label: String) {
+        self.labels.remove(&(self.root_id, label));
+    }
+}
+
+impl Visitor for ExtraLabelVisitor {
```
Comment (leops) [inline crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs:97]: It seems unnecessary to use a custom visitor for this, instead the rule could simply match on `JsBreakStatement` and `JsContinueStatement` with a label and visit the parent nodes to find the corresponding `JsLabeledStatement` and collect all the `JsBreakableStatement` nodes in-between

**Comment by Conaclos** [comment]
There is an issue with the PR workflow...

**Comment by ematipico** [comment]
> There is an issue with the PR workflow...

You have to run the command lintdoc again

**Comment by Conaclos** [comment]
> You have to run the command lintdoc again

In fact the issue comes from out-dated base. This is due to the merge of #4099. All PR introducing a new rule has to be rebased on current main.

File: crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            AnyJsStatement::JsContinueStatement(x) => (x.continue_token().ok()?, x.label_token()?),
+            _ => return None,
+        };
+        let mut trailing_trivia = kw_token
+            .trailing_trivia()
+            .pieces()
+            .rev()
+            .skip_while(|p| p.is_newline() || p.is_whitespace())
+            .collect::<Vec<_>>();
+        trailing_trivia.reverse();
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs:117]: Would you mind adding a comment explaining why we need to apply the "reverse" twice?

File: crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            label_token.text_trimmed_range(),
+            markup! {
+                "Unnecessary "<Emphasis>"label"</Emphasis>"."
+            },
+        ))
+    }
+
+    fn action(ctx: &RuleContext<Self>, _: &Self::State) -> Option<JsRuleAction> {
+        let stmt = ctx.query();
+        let (kw_token, label_token) = match stmt {
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs:106]: Nit: what does "kw" stand for? Maybe we can choose a better word

File: crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        let label_token = match stmt {
+            AnyJsStatement::JsBreakStatement(x) => x.label_token(),
+            AnyJsStatement::JsContinueStatement(x) => x.label_token(),
+            _ => None,
+        }?;
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            label_token.text_trimmed_range(),
+            markup! {
+                "Unnecessary "<Emphasis>"label"</Emphasis>"."
```
Comment (ematipico) [inline crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs:99]: We should at least explain briefly why the label is unnecessary

**Review by ematipico — APPROVED** [review]

File: crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs (hunk 3 above)
Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs:106]: This is an abbreviation of `KeyWord`. However, I can rename the variable to `stmt_token`.

File: crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs (hunk 2 above)
Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs:117]: The intent was to remove trailing spaces until a comment is met. However, I am no longer sure to see the value of that. I could simply remove any trailing trivia?

Comment (ematipico) [reply crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs]: I don't mind seeing a KISS implementation for now. We can always improve the rule later

Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs]: I improved trivia handling. I explain my decision in the comments.
Furthermore, I also rebased the PR.
I think it is now ready for merging :)

Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/no_extra_labels.rs]: I improved trivia handling. I explain my decision in the comments.
Furthermore, I also rebased the PR.
I think it is now ready for merging :)

**Pull request merged**